Fix progress bar float input bug

notiona opened the pull request:
<!--
⚠️ BEFORE CONTRIBUTING PLEASE READ OUR CONTRIBUTING GUIDELINES!
https://github.com/streamlit/streamlit/wiki/Contributing
-->

## Describe your changes
This PR solves the wrongly raised `streamlit.errors.StreamlitAPIException` when giving a value slightly greater than 1.0 for `st.progress_bar`. The issue is stated [here](https://github.com/streamlit/streamlit/issues/5517).

Even after mitigating the issue directly, the PR also handles cases where given value for `st.progress_bar` is slightly below 0.0, which should also cause a similar bug.

Instead of simply rounding the given float value, I have decided to use `math.isclose()` function as it provides both relative tolerance and absolute tolerance capabilities. Also, performance-wise, by checking `math.isclose()` in the same if condition as the previous condition, we would have no excess computation for normal values due to short circuit evaluation.

Additionally if required, this helper function `_check_float_between` can be reused in other parts of streamlit where flexible float comparison is necessary. Thus the low and high bounds are configurable as keyword arguments.

## GitHub Issue Link (if applicable)
Closes https://github.com/streamlit/streamlit/issues/5517

## Testing Plan
- I have made an additional unit test in `lib/tests/streamlit/elements/progress_test.py` checking for values that are off the bound of `(low <= value <= high)`, yet still should be considered valid input.
- I have also tested the case stated in https://github.com/streamlit/streamlit/issues/5517 locally (following the local setup guide in the [contribution guides](https://github.com/streamlit/streamlit/wiki/Contributing)), and it works as intended.

![image](https://github.com/streamlit/streamlit/assets/37995888/63272a85-fbe1-4287-94ed-f12ed710312a)

Please feel free to give any ideas on improvements or suggestions. 😄 

---


**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

notiona committed:
Fix progress bar float input bug

sfc-gh-lmasuch reviewed (approved):
Thanks for the contribution 👍  LGTM

lukasmasuch reviewed (approved):
Thanks for the contribution 👍  LGTM

raethlein commented on lib/streamlit/elements/progress.py:
```diff
@@ -29,6 +30,27 @@
+def _check_float_between(value: float, low: float = 0.0, high: float = 1.0) -> bool:
+    """
+    Checks given value is 'between' the bounds of [low, high],
+    considering close values around bounds are acceptable input
```
I think it would be great if you extended the comment slightly explaining *why* we want to consider close values around bounds, e.g. `... acceptable input, for example in order to account for values being slightly off due to floating point operations.`. What do you think?

notiona replied:
@raethlein I agree. Thank you for your comment. I have extended the docstring in commit [98ff93b4b3af986ef4fde89b8689a22c81dc7a09](https://github.com/streamlit/streamlit/pull/7953/commits/98ff93b4b3af986ef4fde89b8689a22c81dc7a09).

raethlein reviewed (approved):
Thank you, looks great! I have just left a tiny nit.

vdonato commented on lib/streamlit/elements/progress.py:
```diff
@@ -29,6 +30,27 @@
+    low : float
+    high : float
+    """
+    if (
+        (low <= value <= high)
+        or math.isclose(value, low, rel_tol=1e-9, abs_tol=1e-9)
+        or math.isclose(value, high, rel_tol=1e-9, abs_tol=1e-9)
+    ):
+        return True
+    else:
+        return False
```
nit: this could be cleaned up a tiny bit by just returning the conditional in the `if` rather than returning `True` or `False` based on it

```python3
    return (
        (low <= value <= high)
        or math.isclose(value, low, rel_tol=1e-9, abs_tol=1e-9)
        or math.isclose(value, high, rel_tol=1e-9, abs_tol=1e-9)
    )
```

notiona replied:
@vdonato I thought at first since the if condition is quite long it would hurt readability to simply return the expression, but now that I see your suggestion it looks cleaner. Thank you for the comment. I have fixed it in commit [98ff93b4b3af986ef4fde89b8689a22c81dc7a09](https://github.com/streamlit/streamlit/pull/7953/commits/98ff93b4b3af986ef4fde89b8689a22c81dc7a09).

vdonato reviewed (approved):
One tiny nit (and also seconding @raethlein's comment), but otherwise this LGTM. Thanks for the contribution @notiona!

notiona commented:
Thank you for your valuable comments guys! I will address them and let you guys know when I am done.

notiona committed:
add docstring, simplify fix st.progress bar bug fix

notiona commented:
I have addressed all your comments. Feel free to provide any other feedback. Thank you for the great work! @LukasMasuch @sfc-gh-lmasuch  @raethlein @vdonato

lukasmasuch commented:
Thanks for the update :)
